common/src/main/java/tachyon/exception/TachyonException.java

```diff
@@ -15,15 +15,64 @@
 
 package tachyon.exception;
 
+import tachyon.thrift.TachyonTException;
+
+/**
+ * General TachyonException used throughout the system. It must be able serialize itself to the RPC
+ * framework and convert back without losing any necessary information.
+ */
 public class TachyonException extends Exception {
   private TachyonExceptionType mType;
 
-  public TachyonException(Exception e, TachyonExceptionType type) {
-    super(e.getMessage());
+  public TachyonException(TachyonTException te) {
+    super(te.getMessage());
+    mType = TachyonExceptionType.valueOf(te.type);
+  }
+
+  protected TachyonException(TachyonExceptionType type, Throwable cause) {
+    super(cause);
+    mType = type;
+  }
+
+  protected TachyonException(TachyonExceptionType type, String message) {
+    super(message);
+    mType = type;
+  }
+
+  protected TachyonException(TachyonExceptionType type, String message, Throwable cause) {
+    super(message, cause);
     mType = type;
   }
 
   public TachyonExceptionType getType() {
     return mType;
   }
+
+  public TachyonTException toTachyonTException() {
+    return new TachyonTException(mType.name(), getMessage());
+  }
+
+  public void unwrapFileAlreadyExistsException() throws FileAlreadyExistsException {
```

Comment:
I am not sure if this can be done in Java, but would it be possible to instead create a utility class that has a static method:

``` 
public static void unwrap<T>(TachyonException e) throws T {
  if (e.getType == T.getType()) {
    throw new T(e.getMessage(), e);
  }
}
```

My goal is to have a single definition for this repetitive code. Ideally, the use would be for instance ``TachyonException.unwrap<InvalidPathException>(e)``.